### ISSUE TITLE
Constructor for DataView is too broad

### ISSUE THREAD

**Issue by ngbrown** [microsoft/TypeScript#52815]
# lib Update Request

## Configuration Check

My compilation *target* is `ES2021` and my *lib* is `["DOM", "DOM.Iterable", "ES2021"]`. TypeScript version is **4.9.5**.

## Missing / Incorrect Definition

The `DataView` constructor parameter type is incorrectly broad by accepting `ArrayBufferLike` because `TypedArray`'s such as `Uint8Array` match the `ArrayBufferLike` type but are not valid parameters. I think the `DataView` constructor parameter should only accept `ArrayBuffer | SharedArrayBuffer`.

## Sample Code

The following produces no error in TypeScript but does produce an error in Node 16, Chrome 109, and FireFox 111b1:

```js
new DataView(new Uint8Array(32));
```

The runtime error is:

```
Uncaught TypeError: First argument to DataView constructor must be an ArrayBuffer
    at new DataView (<anonymous>)
```

The correct code that does run is:

```js
new DataView(new Uint8Array(32).buffer)
```

## Documentation Link

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView/DataView
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/buffer
- https://tc39.es/ecma262/multipage/structured-data.html#sec-dataview-constructor (specifically a `TypeError` is to be thrown if this is a [detached buffer](https://tc39.es/ecma262/multipage/structured-data.html#sec-isdetachedbuffer), i.e. without array buffer data)

**Comment by fatcerberus** [comment]
`ArrayBufferLike` is already defined as `ArrayBuffer | SharedArrayBuffer`.  The problem is that TS is structurally typed and typed arrays are structurally compatible with `ArrayBuffer`:

```ts
const foo: ArrayBuffer = new Uint8Array(42);  // no error
```

**Comment by ngbrown** [comment]
I knew that to a degree (duck typing and all), and I suppose someone will declare this as duplicate of #202, which I didn't know about.  But this seems like these types should be the exception to structural typing because the browser is much more strict about the types actually being `instanceof`.

**Comment by fatcerberus** [comment]
Yeah, I think there's a feature request somewhere to be able to expand types in function hints.

Apparently this is intentional because doing otherwise is a breaking change: https://github.com/microsoft/TypeScript/issues/31311#issuecomment-490690695

**Comment by ngbrown** [comment]
Interestingly, `never` and `Exclude<T, U>` can almost help:

```ts
interface DataViewConstructor {
    readonly prototype: DataView;
    new<T extends (ArrayBuffer | SharedArrayBuffer) = ArrayBufferLike>(buffer: T & Exclude<T, {BYTES_PER_ELEMENT: number}>, byteOffset?: number, byteLength?: number): DataView;
}
declare var DataView: DataViewConstructor;
```

But it is a breaking change for types that inherit from `DataView`... Maybe someone smarter can make it work right without a generic?

**Comment by RyanCavanaugh** [comment]
I think this is actually a lot easier to accomplish:
```ts
new(buffer: ArrayBufferLike & { BYTES_PER_ELEMENT?: never }, byteOffset?: number, byteLength?: number): DataView;
```
